File: src/Contract/Datamap.fs
namespace ARCtrl.Contract

open ARCtrl.FileSystem
open ARCtrl.ArcPathHelper
open ARCtrl.Spreadsheet
open ARCtrl
open ARCtrl.Helper
open FsSpreadsheet

[<AutoOpen>]
module DatamapContractExtensions = 

    let (|DatamapPath|_|) (input) =
        match input with
        | [|AssaysFolderName; anyAssayName; DatamapFileName|] -> 
            let path = ARCtrl.ArcPathHelper.combineMany input
            Some path
        | [|StudiesFolderName; anyStudyName; DatamapFileName|] -> 
            let path = ARCtrl.ArcPathHelper.combineMany input
            Some path
        | [|WorkflowsFolderName; anyWorkflowName; DatamapFileName|] ->
            let path = ARCtrl.ArcPathHelper.combineMany input
            Some path
        | [|RunsFolderName; anyRunName; DatamapFileName|] ->
            let path = ARCtrl.ArcPathHelper.combineMany input
            Some path
        | _ -> None

    type Datamap with

        // Assay

        member this.ToCreateContractForAssay (assayIdentifier : string) =
            let path = Identifier.Assay.datamapFileNameFromIdentifier assayIdentifier
            Contract.createCreate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))

        member this.ToUpdateContractForAssay (assayIdentifier : string) =
            let path = Identifier.Assay.datamapFileNameFromIdentifier assayIdentifier
            let c = Contract.createUpdate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))
            c

        member this.ToDeleteContractForAssay (assayIdentifier : string) =
            let path = Identifier.Assay.datamapFileNameFromIdentifier assayIdentifier
            let c = Contract.createDelete(path)
            c

        static member toDeleteContractForAssay (assayIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToDeleteContractForAssay(assayIdentifier)

        static member toUpdateContractForAssay (assayIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToUpdateContractForAssay(assayIdentifier)

        static member tryFromReadContractForAssay (assayIdentifier : string) (c:Contract) =
            let path = Identifier.Assay.datamapFileNameFromIdentifier assayIdentifier
            match c with
            | {Path = p; Operation = READ; DTOType = Some DTOType.ISA_Datamap; DTO = Some (DTO.Spreadsheet fsworkbook)} when p = path ->
                let dm = 
                    fsworkbook :?> FsWorkbook
                    |> Datamap.fromFsWorkbook                
                dm.StaticHash <- dm.GetHashCode()
                dm
                |> Some 
            | _ -> None

        // Study

        member this.ToCreateContractForStudy (studyIdentifier : string) =
            let path = Identifier.Study.datamapFileNameFromIdentifier studyIdentifier
            Contract.createCreate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))

        member this.ToUpdateContractForStudy (studyIdentifier : string) =
            let path = Identifier.Study.datamapFileNameFromIdentifier studyIdentifier
            let c = Contract.createUpdate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))
            c

        member this.ToDeleteContractForStudy (studyIdentifier : string) =
            let path = Identifier.Study.datamapFileNameFromIdentifier studyIdentifier
            let c = Contract.createDelete(path)
            c

        static member toDeleteContractForStudy (studyIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToDeleteContractForStudy(studyIdentifier)

        static member toUpdateContractForStudy (studyIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToUpdateContractForStudy(studyIdentifier)


        static member tryFromReadContractForStudy (studyIdentifier : string) (c:Contract) =
            let path = Identifier.Study.datamapFileNameFromIdentifier studyIdentifier
            match c with
            | {Path = p; Operation = READ; DTOType = Some DTOType.ISA_Datamap; DTO = Some (DTO.Spreadsheet fsworkbook)} when p = path->
                let dm = 
                    fsworkbook :?> FsWorkbook
                    |> Datamap.fromFsWorkbook                
                dm.StaticHash <- dm.GetHashCode()
                Some (dm)
            | _ -> None

        // Workflow

        member this.ToCreateContractForWorkflow (workflowIdentifier : string) =
            let path = Identifier.Workflow.datamapFileNameFromIdentifier workflowIdentifier
            Contract.createCreate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))

        member this.ToUpdateContractForWorkflow (workflowIdentifier : string) =
            let path = Identifier.Workflow.datamapFileNameFromIdentifier workflowIdentifier
            let c = Contract.createUpdate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))
            c

        member this.ToDeleteContractForWorkflow (workflowIdentifier : string) =
            let path = Identifier.Workflow.datamapFileNameFromIdentifier workflowIdentifier
            let c = Contract.createDelete(path)
            c

        static member toDeleteContractForWorkflow (workflowIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToDeleteContractForWorkflow(workflowIdentifier)

        static member toUpdateContractForWorkflow (workflowIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToUpdateContractForWorkflow(workflowIdentifier)

        static member tryFromReadContractForWorkflow (workflowIdentifier : string) (c:Contract) =
            let path = Identifier.Workflow.datamapFileNameFromIdentifier workflowIdentifier
            match c with
            | {Path = p; Operation = READ; DTOType = Some DTOType.ISA_Datamap; DTO = Some (DTO.Spreadsheet fsworkbook)} when p = path->
                let dm = 
                    fsworkbook :?> FsWorkbook
                    |> Datamap.fromFsWorkbook                
                dm.StaticHash <- dm.GetHashCode()
                Some (dm)
            | _ -> None

        // Run

        member this.ToCreateContractForRun (runIdentifier : string) =
            let path = Identifier.Run.datamapFileNameFromIdentifier runIdentifier
            Contract.createCreate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))

        member this.ToUpdateContractForRun (runIdentifier : string) =
            let path = Identifier.Run.datamapFileNameFromIdentifier runIdentifier
            let c = Contract.createUpdate(path, DTOType.ISA_Datamap, DTO.Spreadsheet (this |> Datamap.toFsWorkbook))
            c

        member this.ToDeleteContractForRun (runIdentifier : string) =
            let path = Identifier.Run.datamapFileNameFromIdentifier runIdentifier
            let c = Contract.createDelete(path)
            c

        static member toDeleteContractForRun (runIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToDeleteContractForRun(runIdentifier)

        static member toUpdateContractForRun (runIdentifier : string) =
            fun (datamap : Datamap) -> 
                datamap.ToUpdateContractForRun(runIdentifier)

        static member tryFromReadContractForRun (runIdentifier : string) (c:Contract) =
            let path = Identifier.Run.datamapFileNameFromIdentifier runIdentifier
            match c with
            | {Path = p; Operation = READ; DTOType = Some DTOType.ISA_Datamap; DTO = Some (DTO.Spreadsheet fsworkbook)} when p = path->
                let dm = 
                    fsworkbook :?> FsWorkbook
                    |> Datamap.fromFsWorkbook                
                dm.StaticHash <- dm.GetHashCode()
                Some (dm)
            | _ -> None

File: js/Contract/Datamap.js
import { equalsWith, item } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { combineMany } from "../FileSystem/Path.js";
import { safeHash, defaultOf } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { Study_datamapFileNameFromIdentifier, Assay_datamapFileNameFromIdentifier } from "../Core/Helper/Identifier.js";
import { Contract } from "./Contract.js";
import { fromFsWorkbook, toFsWorkbook } from "../Spreadsheet/DataMap.js";
import { DataMap__set_StaticHash_Z524259A4 } from "../Core/DataMap.js";

export function $007CDatamapPath$007C_$007C(input) {
    let matchResult;
    if (!equalsWith((x, y) => (x === y), input, defaultOf()) && (input.length === 3)) {
        if (item(0, input) === "assays") {
            if (item(2, input) === "isa.datamap.xlsx") {
                matchResult = 0;
            }
            else {
                matchResult = 2;
            }
        }
        else if (item(0, input) === "studies") {
            if (item(2, input) === "isa.datamap.xlsx") {
                matchResult = 1;
            }
            else {
                matchResult = 2;
            }
        }
        else {
            matchResult = 2;
        }
    }
    else {
        matchResult = 2;
    }
    switch (matchResult) {
        case 0: {
            const anyAssayName = item(1, input);
            return combineMany(input);
        }
        case 1: {
            const anyStudyName = item(1, input);
            return combineMany(input);
        }
        default:
            return undefined;
    }
}

export function ARCtrl_DataMap__DataMap_ToCreateContractForAssay_Z721C83C5(this$, assayIdentifier) {
    const path = Assay_datamapFileNameFromIdentifier(assayIdentifier);
    return Contract.createCreate(path, "ISA_Datamap", toFsWorkbook(this$));
}

export function ARCtrl_DataMap__DataMap_ToUpdateContractForAssay_Z721C83C5(this$, assayIdentifier) {
    const path = Assay_datamapFileNameFromIdentifier(assayIdentifier);
    return Contract.createUpdate(path, "ISA_Datamap", toFsWorkbook(this$));
}

export function ARCtrl_DataMap__DataMap_ToDeleteContractForAssay_Z721C83C5(this$, assayIdentifier) {
    const path = Assay_datamapFileNameFromIdentifier(assayIdentifier);
    return Contract.createDelete(path);
}

export function ARCtrl_DataMap__DataMap_toDeleteContractForAssay_Static_Z721C83C5(assayIdentifier) {
    return (dataMap) => ARCtrl_DataMap__DataMap_ToDeleteContractForAssay_Z721C83C5(dataMap, assayIdentifier);
}

export function ARCtrl_DataMap__DataMap_toUpdateContractForAssay_Static_Z721C83C5(assayIdentifier) {
    return (dataMap) => ARCtrl_DataMap__DataMap_ToUpdateContractForAssay_Z721C83C5(dataMap, assayIdentifier);
}

export function ARCtrl_DataMap__DataMap_tryFromReadContractForAssay_Static(assayIdentifier, c) {
    let fsworkbook;
    const path = Assay_datamapFileNameFromIdentifier(assayIdentifier);
    let matchResult, fsworkbook_1, p_1;
    if (c.Operation === "READ") {
        if (c.DTOType != null) {
            if (c.DTOType === "ISA_Datamap") {
                if (c.DTO != null) {
                    if ((fsworkbook = c.DTO, c.Path === path)) {
                        matchResult = 0;
                        fsworkbook_1 = c.DTO;
                        p_1 = c.Path;
                    }
                    else {
                        matchResult = 1;
                    }
                }
                else {
                    matchResult = 1;
                }
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
    }
    else {
        matchResult = 1;
    }
    switch (matchResult) {
        case 0: {
            const dm = fromFsWorkbook(fsworkbook_1);
            DataMap__set_StaticHash_Z524259A4(dm, safeHash(dm));
            return dm;
        }
        default:
            return undefined;
    }
}

export function ARCtrl_DataMap__DataMap_ToCreateContractForStudy_Z721C83C5(this$, studyIdentifier) {
    const path = Study_datamapFileNameFromIdentifier(studyIdentifier);
    return Contract.createCreate(path, "ISA_Datamap", toFsWorkbook(this$));
}

export function ARCtrl_DataMap__DataMap_ToUpdateContractForStudy_Z721C83C5(this$, studyIdentifier) {
    const path = Study_datamapFileNameFromIdentifier(studyIdentifier);
    return Contract.createUpdate(path, "ISA_Datamap", toFsWorkbook(this$));
}

export function ARCtrl_DataMap__DataMap_ToDeleteContractForStudy_Z721C83C5(this$, studyIdentifier) {
    const path = Study_datamapFileNameFromIdentifier(studyIdentifier);
    return Contract.createDelete(path);
}

export function ARCtrl_DataMap__DataMap_toDeleteContractForStudy_Static_Z721C83C5(studyIdentifier) {
    return (dataMap) => ARCtrl_DataMap__DataMap_ToDeleteContractForStudy_Z721C83C5(dataMap, studyIdentifier);
}

export function ARCtrl_DataMap__DataMap_toUpdateContractForStudy_Static_Z721C83C5(studyIdentifier) {
    return (dataMap) => ARCtrl_DataMap__DataMap_ToUpdateContractForStudy_Z721C83C5(dataMap, studyIdentifier);
}

export function ARCtrl_DataMap__DataMap_tryFromReadContractForStudy_Static(studyIdentifier, c) {
    let fsworkbook;
    const path = Study_datamapFileNameFromIdentifier(studyIdentifier);
    let matchResult, fsworkbook_1, p_1;
    if (c.Operation === "READ") {
        if (c.DTOType != null) {
            if (c.DTOType === "ISA_Datamap") {
                if (c.DTO != null) {
                    if ((fsworkbook = c.DTO, c.Path === path)) {
                        matchResult = 0;
                        fsworkbook_1 = c.DTO;
                        p_1 = c.Path;
                    }
                    else {
                        matchResult = 1;
                    }
                }
                else {
                    matchResult = 1;
                }
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
    }
    else {
        matchResult = 1;
    }
    switch (matchResult) {
        case 0: {
            const dm = fromFsWorkbook(fsworkbook_1);
            DataMap__set_StaticHash_Z524259A4(dm, safeHash(dm));
            return dm;
        }
        default:
            return undefined;
    }
}

